compiler/rustc_parse/src/parser/attr_wrapper.rs

```diff
@@ -400,24 +400,25 @@ fn make_token_stream(
 ) -> AttrAnnotatedTokenStream {
     #[derive(Debug)]
     struct FrameData {
-        open: Span,
-        open_delim: DelimToken,
+        open_delim_sp: Option<(DelimToken, Span)>,
         inner: Vec<(AttrAnnotatedTokenTree, Spacing)>,
     }
-    let mut stack =
-        vec![FrameData { open: DUMMY_SP, open_delim: DelimToken::NoDelim, inner: vec![] }];
+    let mut stack = vec![FrameData { open_delim_sp: None, inner: vec![] }];
     let mut token_and_spacing = iter.next();
     while let Some((token, spacing)) = token_and_spacing {
         match token {
             FlatToken::Token(Token { kind: TokenKind::OpenDelim(delim), span }) => {
-                stack.push(FrameData { open: span, open_delim: delim, inner: vec![] });
+                stack.push(FrameData { open_delim_sp: Some((delim, span)), inner: vec![] });
             }
             FlatToken::Token(Token { kind: TokenKind::CloseDelim(delim), span }) => {
                 // HACK: If we encounter a mismatched `None` delimiter at the top
                 // level, just ignore it.
```

Comment:
Is this still necessary?
There's no longer a `DelimToken::NoDelim` at the top level.

Reply:
There's also a similar `// HACK` comment lower in the same function.

Reply:
The "mismatched `None` delimiter at the top level" here refers to `token`, which comes from the iterator, not the frame. Indeed, the condition looks for a non-`NoDelim` in the frame. So this change preserves the existing behaviour, as I understand it.

More generally: there are three `HACK` comments in this function, and a comment at the top of the function referring to #67062. The `HACK` comments are on three `if` statements. I tried changing the bodies of those three `if` statements to `panic!()` and the test suite passed. I then tried removing all three `if` statements and the test suite again passed.

These `if` statements were added in #82608, with an explanation [here](https://github.com/rust-lang/rust/pull/82608#issuecomment-812877329). I don't know if (a) they never did anything useful, (b) they do something useful but we have no test coverage for them, or (c) something has changed in the meantime that means they are no longer necessary.

I suggest we leave this change as is, and consider removing these hacks in a separate PR. That way if there are any regressions they'll be easier to deal with.

Reply:
I have filed #96543 to remove the hacks.